File: leetcode/15_backtracking/78_Subsets.go
package main

import "fmt"

func subsets(nums []int) [][]int {
	if nums == nil {
		return [][]int{}
	}

	result := [][]int{}
	backtrack(nums, &result, []int{}, 0)
	return result
}

func backtrack(nums []int, result *[][]int, path []int, start int) {
	subsetCopy := make([]int, len(path))
	copy(subsetCopy, path)
	*result = append(*result, subsetCopy)

	for i := start; i < len(nums); i++ {
		path = append(path, nums[i])
		backtrack(nums, result, path, i+1)
		path = path[:len(path)-1]
	}
}

func main() {
	testCase1 := []int{1, 2, 3}
	testCase2 := []int{1, 2}
	testCase3 := []int{0}

	fmt.Println(subsets(testCase1))
	fmt.Println(subsets(testCase2))
	fmt.Println(subsets(testCase3))
}

File: leetcode/15_backtracking/78_Subsets.py
class Solution:
    def subsets(self, nums: list[int]) -> list[list[int]]:
        if not nums:
            return [[]]

        result = []
        self.backtrack(nums, result, [], 0)
        return result

    def backtrack(self, nums, result, path, start):
        subset_copy = path[:]
        result.append(subset_copy)

        for i in range(start, len(nums)):
            path.append(nums[i])
            self.backtrack(nums, result, path, i + 1)
            path.pop()


if __name__ == '__main__':
    test_case1 = [1, 2, 3]
    test_case2 = [1, 2]
    test_case3 = [0]

    print(Solution().subsets(test_case1))
    print(Solution().subsets(test_case2))
    print(Solution().subsets(test_case3))
